src/cls/rbd/cls_rbd.cc

```diff
@@ -3599,6 +3604,65 @@ int image_status_remove_down(cls_method_context_t hctx) {
   return 0;
 }
 
+int instances_list(cls_method_context_t hctx,
+                   map<std::string, cls::rbd::MirrorInstanceState> *instances) {
+  std::string last_read = INSTANCE_KEY_PREFIX;
+  int max_read = RBD_MAX_KEYS_READ;
+  int r = max_read;
+  while (r == max_read) {
+    std::map<std::string, bufferlist> vals;
+    r = cls_cxx_map_get_vals(hctx, last_read, INSTANCE_KEY_PREFIX.c_str(),
+                             max_read, &vals);
+    if (r < 0) {
+      CLS_ERR("error reading mirror instances: %s", cpp_strerror(r).c_str());
+      return r;
+    }
+
+    for (auto &it : vals) {
+      const std::string &id = it.first.substr(INSTANCE_KEY_PREFIX.size());
+      try {
+        bufferlist::iterator bl_it = it.second.begin();
+        cls::rbd::MirrorInstanceState state;
+	::decode(state, bl_it);
+        (*instances)[id] = state;
+      } catch (const buffer::error &err) {
+	CLS_ERR("could not decode mirror instance '%s'", it.first.c_str());
+	return -EIO;
+      }
+    }
+
+    if (!vals.empty()) {
+      last_read = vals.rbegin()->first;
+    }
+  }
+  return 0;
+}
+
+int instances_set(cls_method_context_t hctx, const string &instance_id,
+                  const cls::rbd::MirrorInstanceState &state) {
+  bufferlist bl;
+  ::encode(state, bl);
+
+  int r = cls_cxx_map_set_val(hctx, instance_key(instance_id), &bl);
```

Comment:
Actually, I was thinking about providing functions like below instead of instances_set:

- instances_add/register -- add a record in up state
- instances_down/pruning -- change a record state to down
- instances_remove/unregister -- remove a record

I can do any way: either add a check to instances_set or introduce add/down functions instead of set. What do you like more?